File: R/tutorial/simple-examples/foreach_flow.R
library(metaflow)

start <- function(self){
    self$params= c("param1", "param2", "param3") 
}

a <- function(self){
    self$result = paste(self$input, "processed")
}

join <- function(self, inputs){
    results <- gather_inputs(inputs, "result")
    print(results)
}

metaflow("ForeachFlow") %>%
    step(step = "start",
         r_function = start,
         next_step = "a",
         foreach="params") %>%
    step(step = "a", 
         r_function=a, 
         next_step="join") %>%
    step(step="join", 
         r_function=join, 
         next_step="end",
         join=TRUE) %>%
    step(step="end") %>%
    run()

File: test/ux/core/flows/dag/foreach_flow.py
from metaflow import FlowSpec, step, project


@project(name="foreach_flow")
class ForeachFlow(FlowSpec):
    @step
    def start(self):
        self.items = [1, 2, 3, 4, 5]
        self.next(self.process, foreach="items")

    @step
    def process(self):
        self.result = self.input * 2
        self.next(self.join)

    @step
    def join(self, inputs):
        self.results = sorted([i.result for i in inputs])
        self.next(self.end)

    @step
    def end(self):
        pass


if __name__ == "__main__":
    ForeachFlow()
